jetty/src/main/scala/com/avsystem/commons/jetty/rpc/JettyRPCFramework.scala

```diff
@@ -89,30 +88,38 @@ object JettyRPCFramework extends StandardRPCFramework with LazyLogging {
       request(HttpMethod.PUT, call)
   }
 
-  class RPCHandler(rootRpc: RawRPC, contextTimeout: FiniteDuration) extends AbstractHandler {
-    override def handle(target: String, baseRequest: Request, request: HttpServletRequest, response: HttpServletResponse): Unit = {
-      baseRequest.setHandled(true)
-
-      val content = Iterator.continually(request.getReader.readLine())
-        .takeWhile(_ != null)
-        .mkString("\n")
-
-      val call = read[Call](new RawValue(content))
+  class RPCHandler(rootRpc: RawRPC, contextTimeout: FiniteDuration) extends HttpServlet {
+    override def service(request: HttpServletRequest, response: HttpServletResponse): Unit = {
+      // readRequest must execute in request thread but we want exceptions to be handled uniformly, hence the Try
+      val content =
+        Using(request.getReader)(reader =>
+          Iterator.continually(reader.readLine()).takeWhile(_ != null).mkString("\n")
+        )
+      val call = content.map(content => read[Call](new RawValue(content)))
 
       HttpMethod.fromString(request.getMethod) match {
         case HttpMethod.POST =>
-          val async = request.startAsync().setup(_.setTimeout(contextTimeout.toMillis))
-          handlePost(call).andThenNow {
+          val asyncContext = request.startAsync().setup(_.setTimeout(contextTimeout.toMillis))
+          val completed = new AtomicBoolean(false)
+          // Need to protect asyncContext from being completed twice because after a timeout the
+          // servlet may recycle the same context instance between subsequent requests (not cool)
+          // https://stackoverflow.com/a/27744537
+          def completeWith(code: => Unit): Unit =
+            if (!completed.getAndSet(true)) {
+              code
+              asyncContext.complete()
+            }
+          completeWith(Future.fromTry(call).flatMapNow(handlePost).andThenNow {
             case Success(responseContent) =>
               response.setContentType(MimeTypes.Type.APPLICATION_JSON.asString())
               response.setCharacterEncoding(StandardCharsets.UTF_8.name())
               response.getWriter.write(responseContent.s)
             case Failure(t) =>
               response.sendError(HttpStatus.INTERNAL_SERVER_ERROR_500, t.getMessage)
               logger.error("Failed to handle RPC call", t)
-          }.andThenNow { case _ => async.complete() }
+          })
```

Comment:
That does not seem right to me, `asyncContext.complete()` is not called after `Future`. Argument `code` in `completeWith` method is a Future, so:
```
code // some future running in background
asyncContext.complete() // imidietly closes context without waiting for actual completion

```